qt-models/cylindermodel.cpp

```diff
@@ -602,6 +615,46 @@ void CylindersModel::moveAtFirst(int cylid)
 	endMoveRows();
 }
 
+void CylindersModel::moveAtPos(int cylid, int pos)
+{
+	int mapping[MAX_CYLINDERS];
+	cylinder_t temp_cyl;
+
+	if (cylid == pos)
+		return;
+
+	beginMoveRows(QModelIndex(), cylid, cylid, QModelIndex(), pos + 1);
+	memmove(&temp_cyl, &displayed_dive.cylinder[cylid], sizeof(temp_cyl));
```

Comment:
Hmm... In C++ you can simply assign structs: `temp_cyl=displayed_dive.cylinder[cylid]`. Wouldn't that be more readable?

Reply:
By the way, you could shorten the whole reorder logic significantly, but I don't know which of these constructs are acceptable:

```
#include <algorithm>
#include <numeric>
...
int mapping[MAX_CYLINDERS];
std::iota(&mapping[0], &mapping[MAX_CYLINDERS], 0); // Fill with 0..MAX_CYLINDERS-1
auto &cylinder = displayed_dive.cylinder;
if (pos > cylid) {
  std::rotate(&cylinder[cylid], &cylinder[cylid + 1], &cylinder[pos + 1]);
  std::rotate(&mapping[cylid], &mapping[cylid + 1], &mapping[pos + 1]);
} else {
  std::rotate(&cylinder[pos], &cylinder[cylid], &cylinder[cylid + 1]);
  std::rotate(&mapping[pos], &mapping[cylid], &mapping[cylid + 1]);
}
...

```
In my opinion, this is more readable than the various for-loops, but YMMV.

Reply:
I am infamous for not wanting to add too much C++-y-ness to our code. But yeah, wow, that looks much nicer to me :-)
:+1:

Reply:
Doing a quick grep also shows me that we are already using such constructs like std:sort and others in Subsurface code.
So ok for me to change this, but:
- Will take me some time because I want to really learn and understand this, not just copy&paste and I then want to update at least also similar functions in same file (->seperate commits)
- I first like to finally know if all this open cylinder handling changes will get approved (and I'm not in a hurry here)

Reply:
- yes on consciously using C++ things like std::rotate, std::sort, etc
- yes on UNDERSTANDING them before using them
- yes on cleaning up things as you go - @bstoeger is doing such an amazing job of this
- I think I would like to delay merging the open cylinder handling until after 4.7.5, but overall I think this is a good thing to work on as there are clearly cases where the current code is just wrong